bn-apps/memberships-management/membership-service/src/main/kotlin/net/corda/businessnetworks/membership/member/service/MembershipsCache.kt

```diff
@@ -1,33 +1,62 @@
 package net.corda.businessnetworks.membership.member.service
 
-import net.corda.businessnetworks.membership.member.MembershipsListResponse
 import net.corda.businessnetworks.membership.states.MembershipState
 import net.corda.core.contracts.StateAndRef
 import net.corda.core.identity.Party
 import net.corda.core.node.AppServiceHub
 import net.corda.core.node.services.CordaService
 import net.corda.core.serialization.SingletonSerializeAsToken
+import java.lang.IllegalArgumentException
+import java.time.Instant
 import java.util.concurrent.ConcurrentHashMap
-import java.util.concurrent.ConcurrentMap
+
+private typealias MembershipByParty = ConcurrentHashMap<Party, StateAndRef<MembershipState<Any>>>
+private typealias MembershipsByBNO = ConcurrentHashMap<Party, MembershipByParty>
 
 /**
  * A singleton service that holds a cache of memberships
  */
 @CordaService
 class MembershipsCacheHolder(private val appServiceHub : AppServiceHub) : SingletonSerializeAsToken() {
-    var cache : MembershipsCache? = null
+    val cache : MembershipsCache = MembershipsCache()
+}
+
+class MembershipsCache {
+    private val cache = MembershipsByBNO()
+    private val lastRefreshed = ConcurrentHashMap<Party, Instant>()
+
+    fun getMembership(bno : Party, party : Party) : StateAndRef<MembershipState<Any>>? = cache[bno]?.get(party)
+
+    fun getMemberships(bno : Party) : MembershipByParty = cache[bno] ?: MembershipByParty()
 
-    fun resetCache() {
-        cache = null
+    fun updateMembership(changedMembership : StateAndRef<MembershipState<Any>>) {
+        val membershipsByParty = cache.getOrPut(changedMembership.state.data.bno) {
+            MembershipByParty()
+        }
+        membershipsByParty.merge(changedMembership.state.data.member, changedMembership) { oldState, newState ->
+            if (oldState.state.data.modified > newState.state.data.modified) oldState else newState
+        }
     }
-}
 
-class MembershipsCache private constructor(val membershipMap : ConcurrentMap<Party, StateAndRef<MembershipState<Any>>>) {
-    companion object {
-        fun from(membershipResponse : MembershipsListResponse) = MembershipsCache(
-                ConcurrentHashMap(membershipResponse.memberships.map { it.state.data.member to it }.toMap()))
+    fun applyMembershipsSnapshot(membershipByParty : List<StateAndRef<MembershipState<Any>>>) {
+        if (membershipByParty.isNotEmpty()) {
+            val bnos = membershipByParty.asSequence().map { it.state.data.bno }.toSet()
+            if (bnos.size != 1) {
+                throw IllegalArgumentException("All membership states in the snapshot should refer to the same BNO!")
```

Comment:
As per the other comment, it's going to break here if the BNO is a member itself of another business network

Reply:
applyMembershipsSnapshot is invoked against of a list of a snapshot of memberships from a BNO. So even if a node is both BNO and a member to other BN, it shouldn't affect this logic.

Reply:
I meant - if the BNO happens to be a member of some other BN then in the memberships the BNO returns will be their own membership of the other BN and with that also the other BNO making it two distinct BNOs in here. Anyway, this should go away if we fix that other part where the BNO returns memberships in GetMembershipsFlowResponder, which (according to your comment) you are going to :-)

Reply:
Right cool :) I'm about to push the changes in a sec